contracts/Compass.vy

```diff
@@ -219,3 +253,40 @@ def deploy_erc20(_paloma_denom: String[64], _name: String[64], _symbol: String[3
     event_id: uint256 = unsafe_add(self.last_event_id, 1)
     self.last_event_id = event_id
     log ERC20DeployedEvent(_paloma_denom, erc20, _name, _symbol, _decimals, event_id)
+
+@external
+def update_treasury(consensus: Consensus, _feeset_id: uint256, _community_fee: uint256, _security_fee: uint256, _community_wallet: address, _security_wallet: address):
+    assert self.feeset_id < _feeset_id, "Old feeset ID"
+    # check if the supplied current validator set matches the saved checkpoint
+    assert self.last_checkpoint == self.make_checkpoint(consensus.valset), "Incorrect Checkpoint"
+    # signing data is keccak256 hash of abi_encoded update_treasury(feeset_id, community_fee, security_fee, community_wallet, security_wallet)
+    args_hash: bytes32 = keccak256(_abi_encode(_feeset_id, _community_fee, _security_fee, _community_wallet, _security_wallet, method_id=method_id("update_treasury(uint256,uint256,uint256,address,address)")))
+    # check if enough validators signed args_hash
+    self.check_validator_signatures(consensus, args_hash)
+    event_id: uint256 = unsafe_add(self.last_event_id, 1)
+    self.last_event_id = event_id
+    self.feeset_id = _feeset_id
+    self.community_fee = _community_fee
+    self.security_fee = _security_fee
+    self.community_wallet = _community_wallet
+    self.security_wallet = _security_wallet
+    log TreasuryUpdatedEvent(_feeset_id, _community_fee, _security_fee, _community_wallet, _security_wallet, event_id)
+
+@external
+@payable
+@nonreentrant('lock')
+def deposit(paloma_address: bytes32, amount: uint256):
+    if msg.value > amount:
+        send(msg.sender, unsafe_sub(msg.value, amount))
+    else:
+        assert msg.value == amount, "Insufficient deposit"
+    _fee: uint256 = unsafe_div(amount * self.security_fee, DENOMINATOR)
+    send(self.security_wallet, _fee)
+    relayer_fee: uint256 = amount - _fee
+    _fee = unsafe_div(amount * self.community_fee, DENOMINATOR)
+    send(self.community_wallet, _fee)
+    relayer_fee -= _fee
+    _nonce: uint256 = self.deposit_nonce
+    self.fee_balance[paloma_address] = unsafe_add(self.fee_balance[paloma_address], relayer_fee)
+    log Deposited(paloma_address, msg.sender, _nonce, relayer_fee)
+    self.deposit_nonce = unsafe_add(_nonce, 1)
```

Comment:
I think the idea to just move funds to community and security wallet once during deposit is great. At the moment, those fees are defined as percentages of the `relayer_fee`, which is unique to each relayer, meaning a more expensive relayer would generate more community/security fee, whereas a cheaper relayer would generate less of it.

It also means that more expensive messages would generate more `community` and `security` fees, while cheap messages would generate less.

I honestly don't know the reasoning behind that decision, it was made before my time with Volume and I cannot find any background on this. @taariq @verabehr might know more. See https://www.notion.so/volumefi/EPIC-09-26-22-Paloma-Gas-Management-Pigeon-Feed-78319280833e4c7c8ffad368eddb1d91?pvs=4#659fbe58c4f84d56bce0f86afb35839e for background.

Moving the deduction of those fees to the deposit stage changes how they should be defined, and how we as a team should understand and communicate them.

Reply:
This is the most cheap solution for gas. Paying fees in SLC will increase SLC gas cost more. SLC gas cost is already too expensive.

Reply:
@byte-bandit @webelf101 what is the decision that needs to be made here?

Reply:
@taariq my understanding is that 
EITHER, we compass becomes more gas expensive and we keep the `community` and `security` as percentage of the relayer fee 
OR `community` and `security` are paid on deposit, but then they would be independent of the actual relayer fee and would need to be a percentage of the deposit amount